### PR TITLE
Add support for base36 and base2 toString

### PR DESCRIPTION
**Pull request by malcolmocean**
I wanted to use https://github.com/iancmyers/active-user, but it requires bigint which I don't have. Tried to switch to bignum, but bignum didn't support `toString(2)`... so I accidentally spent all day (I haven't written C++ code in like 5 years) adding support for it.

Would be nice if it could get merged in so it's available in `npm`. As it is, I'm going to be requiring it using the custom git repo syntax, but that's a bit ugly.

### CODE REVIEW

**Comment by pnicorelli** [comment]
+1 for this, please add it

File: bignum.cc [hunk 1]
```diff
@@ -431,16 +489,63 @@ NAN_METHOD(BigNum::ToString)
...
-  case 10:
-    to = BN_bn2dec(&bignum->bignum_);
-    break;
-  case 16:
-    to = BN_bn2hex(&bignum->bignum_);
-    break;
-  default:
-    Nan::ThrowError("Invalid base, only 10 and 16 are supported");
-    return;
+  if (base == 2) {
```
Comment (tracker1) [inline bignum.cc]: Why isn't this done as a separate method, similar to the other base types?

**Comment by tracker1** [comment]
I think base2 should work like the other base conversions and be a separate function, not inlined as it is.  It's undue extra complexity in the base conversion method.

File: test/big.js [hunk 2]
```diff
@@ -504,6 +504,29 @@ exports.jacobi = function () {
     assert.equal(b1.jacobi(b2), 1);
 };
 
+exports.base2 = function () {
+  assert.equal(bignum('-1234734').toString(2), '-100101101011100101110');
+  assert.equal(bignum('-2').toString(2), '-10');
```
Comment (tracker1) [inline test/big.js]: Is this the correct behavior?  Should this be the bytes for a signed int32/int64 in the case of negative values?

File: bignum.cc [hunk 3]
```diff
@@ -177,6 +177,64 @@ int BN_jacobi_priv(const BIGNUM *A,const BIGNUM *N,int *jacobi,
...
+  char *p;
+  BIGNUM *t = NULL;
+  BN_ULONG *bn_buff, *lp;
+
+  i = BN_num_bits(a);
+#define BN_BASE36_NUM 5 // log(36)
+
+  num = (i + BN_BASE36_NUM + 1) / BN_BASE36_NUM;
+  bn_buff = (BN_ULONG*) OPENSSL_malloc(num * sizeof(BN_ULONG));
+  buf = (char*) OPENSSL_malloc((num + 2) * sizeof(char));
```
Comment (bnoordhuis) [inline bignum.cc]: `sizeof(char)` is always 1 so multiplying by it is not that useful.

For the allocation on the line above, I'd use `sizeof(*bn_buff)`.  That way it keeps working when the type of `bn_buff` changes.

File: bignum.cc [hunk 4]
```diff
@@ -177,6 +177,64 @@ int BN_jacobi_priv(const BIGNUM *A,const BIGNUM *N,int *jacobi,
...
+
+  if (buf == NULL || bn_buff == NULL) {
+    Nan::ThrowError("could not allocate memory for conversion");
```
Comment (bnoordhuis) [inline bignum.cc]: Leaks `bn_buff` when `buf == NULL`.

File: bignum.cc (hunk 1 above)
Comment (bnoordhuis) [inline bignum.cc]: It seems a little wasteful to special-case when you can just generalize `BN_bn2base36(a)` to `BN_bn2base(a, base)` where `2 <= base <= 36`.

File: test/big.js (hunk 2 above)
Comment (bnoordhuis) [inline test/big.js]: Intuitively, I would expect 2's complement.  That said:

1. What should 2's complement look like for arbitrary precision numbers?
2. It's consistent with e.g. `BN_bn2hex()`.

**Pull request closed without merging**